vllm/v1/worker/gpu_model_runner.py

```diff
@@ -520,16 +543,35 @@ def execute_model(
                 inputs_embeds=inputs_embeds,
             )
         hidden_states = hidden_states[:num_scheduled_tokens]
-        hidden_states = hidden_states[logits_indices]
-        logits = self.model.compute_logits(hidden_states, None)
+        sample_hidden_states = hidden_states[sampling_metadata.sample_indicies]
+        sample_logits = self.model.compute_logits(sample_hidden_states, None)
 
         # Sample the next token and get logprobs if needed.
-        sampling_metadata = self._prepare_sampling(scheduler_output)
         sampler_output = self.model.sample(
-            logits=logits,
+            logits=sample_logits,
             sampling_metadata=sampling_metadata,
         )
 
+        # Compute prompt logprobs if needed.
```

Comment:
Previously I had an implementation that only calls `compute_logits` once. It works as follows:
- construct one big mask rather than N masks
- make the temperature and top_p and top_k objects with `repeat_interleave` such that they have shape N

The problem with this is that it became a bit tricky to pull them apart afterwards. If we do want to purse this optimizaiton, we should do it as a follow up to this PR.